### PR TITLE
fixes #377 - add functions to handle bigdecimal / biginteger

### LINKED ISSUE #377: add functions to handle bigdecimal  / biginteger
* string / long from and to BI / BD

* apoc.number.exact.add(string, string)
* apoc.number.exact.sub(string, string)
* apoc.number.exact.mul(string, string[, prec],[roundingMode])
* apoc.number.exact.div(string, string[, prec],[roundingMode])
* apoc.number.exact.toLong(string,[prec as * 10^prec],[roundingMode]))
* apoc.number.exact.toFloat(string,[prec],[roundingMode])
* apoc.number.exact.toExact(Number)

### PR DESCRIPTION
**Pull request by AngeloBusato**

### CODE REVIEW

File: docs/exact.adoc [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+== Handle BigInteger And BigDecimal
+
+[options="header"]
+|===
+| Statement | Description | Return type
+| RETURN apoc.number.exact.add(stringA,stringB) | return the sum's result of two large numbers | string
+| RETURN apoc.number.exact.sub(stringA,stringB) | return the substraction's of two large numbers | string
+| RETURN apoc.number.exact.mul(stringA,stringB,[prec],[roundingModel] | return the multiplication's result of two large numbers | string
+| RETURN apoc.number.exact.div(stringA,stringB,[prec],[roundingModel]) | return the division's result of two large numbers | string
+| RETURN apoc.number.exact.toLong(string,[prec],[roundingMode]) | return the Long value of a large number | Long
```
Comment (jexp) [inline docs/exact.adoc]: change toLong to toInteger -> also replace mentions of "Long" with "Integer"

File: src/main/java/apoc/number/exact/Exact.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ * @since 17 May 2017
+ */
+public class Exact {
+
+	@UserFunction
+	@Description("apoc.number.exact.add(stringA,stringB) | return the sum's result of two large numbers")
+	public String add(@Name("stringA")String stringA, @Name("stringB")String stringB){
+
+		return new BigDecimal(stringA).add(new BigDecimal(stringB)).toPlainString();
```
Comment (jexp) [inline src/main/java/apoc/number/exact/Exact.java]: handle null values for either parameter by returning null for all procedures

Comment (jexp) [inline src/main/java/apoc/number/exact/Exact.java]: And change description to use `-`, instead of `|` to not conflict with asciidoc-tables

**Comment by DanielBerton** [comment]
@jexp fixed requests

File: src/main/java/apoc/number/exact/Exact.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+	@UserFunction
+	@Description("apoc.number.exact.sub(stringA,stringB) - return the substraction's of two large numbers")
+	public String sub(@Name("stringA")String stringA, @Name("stringB")String stringB){
+		if(stringA == null || stringA.isEmpty() || stringB == null || stringB.isEmpty()) return null;
+		return new BigDecimal(stringA).subtract(new BigDecimal(stringB)).toPlainString();
+	}
+
+	@UserFunction
+	@Description("apoc.number.exact.mul(stringA,stringB,[prec],[roundingModel]) - return the multiplication's result of two large numbers ")
```
Comment (jexp) [inline src/main/java/apoc/number/exact/Exact.java]: if prec and roundingMode are optional, we should have defaultValue in their @Name annotations for all methods, that getMatchContext can react to sensibly. @AngeloBusato

File: src/main/java/apoc/number/exact/Exact.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	private MathContext getMatchContext(Long precision, String roundingMode){
+
+		if (precision == null) {
+			precision = Long.valueOf(0);
+		}
+
+		RoundingMode rm = RoundingMode.HALF_UP;
+		if (!StringUtils.isEmpty(roundingMode)) {
```
Comment (jexp) [inline src/main/java/apoc/number/exact/Exact.java]: null check

File: src/main/java/apoc/number/exact/Exact.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+		return new MathContext(precision.intValue(), rm);
+	}
+
+	private MathContext getMatchContextLong(Long precision, String roundingMode){
+		if (precision == null) {
+			precision = Long.valueOf(0);
+		}
+		else {
+			Double pow = pow(10, precision);
```
Comment (jexp) [inline src/main/java/apoc/number/exact/Exact.java]: should this be documented? Or do we need a precision at all for turning decimals into Longs?

Comment (DanielBerton) [inline src/main/java/apoc/number/exact/Exact.java]: I think so, for example 12.54 with `precision` 2 and `RoundingMode.HALF_UP` we have 13 as result, whereas with `precision` 3 we have 12.
We have to improve the documentation.

**Comment by AngeloBusato** [comment]
@jexp fixed

**Comment by jexp** [comment]
Looks good thanks a lot!!

**Pull request merged**